// @com.silencelaboratories/react-native-recovery-kit/lib/cjs/driveEndpoints.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFileData = exports.getUserIdFromAccessToken = exports.updateFileData = exports.createFile = exports.getFile = exports.getFolder = exports.getFilesAndFolder = exports.createFolder = void 0;
const axios_1 = __importDefault(require("axios"));
const createFolder = async (params) => {
    return await (0, axios_1.default)({
        method: "post",
        url: `https://www.googleapis.com/drive/v2/files`,
        headers: {
            Authorization: "Bearer " + params.token,
        },
        data: {
            title: params.folderName,
            mimeType: "application/vnd.google-apps.folder",
        },
    })
        .then((data) => {
        return { id: data.data.id, name: params.folderName };
    })
        .catch((error) => {
        throw new Error("Error: Create Folder Endpoint");
    });
};
exports.createFolder = createFolder;
const getFilesAndFolder = async (params) => {
    return await axios_1.default
        .request({
        method: "get",
        url: `https://www.googleapis.com/drive/v2/files`,
        headers: {
            Accept: "application/json",
            Authorization: "Bearer " + params.token,
        },
    })
        .then((data) => {
        return data.data.items.map((file) => ({
            id: file.id,
            name: file.title,
            type: file.mimeType === "application/vnd.google-apps.folder"
                ? "folder"
                : "file",
            isTrashed: file.explicitlyTrashed,
        }));
    })
        .catch((error) => {
        throw new Error("Error: Get File And Folder Endpoint");
        // return { error: error };
    });
};
exports.getFilesAndFolder = getFilesAndFolder;
const getFolder = async (params) => {
    return await (0, exports.getFilesAndFolder)(params)
        .then((data) => data.filter((f) => f.isTrashed === false && f.type === "folder"))
        .catch((error) => {
        throw new Error(error);
    });
};
exports.getFolder = getFolder;
const getFile = async (params) => {
    return await (0, exports.getFilesAndFolder)(params)
        .then((data) => data.filter((f) => f.isTrashed === false && f.type === "file"))
        .catch((error) => {
        throw new Error(error);
    });
};
exports.getFile = getFile;
const createFile = async (params) => {
    return await (0, axios_1.default)({
        method: "post",
        url: `https://www.googleapis.com/drive/v3/files`,
        headers: {
            Accept: "application/json",
            Authorization: "Bearer " + params.token,
        },
        data: {
            name: params.fileName,
            ...(params.folderId ? { parents: [params.folderId] } : {}),
            mimeType: "text/plain",
        },
    })
        .then((data) => {
        return { id: data.data.id, name: params.fileName };
    })
        .catch((error) => {
        throw new Error("Error: Create File Endpoint");
    });
};
exports.createFile = createFile;
const updateFileData = async (params) => {
    return await (0, axios_1.default)({
        method: "put",
        url: `https://www.googleapis.com/upload/drive/v2/files/${params.fileId}`,
        headers: {
            "Content-Type": "application/json",
            Authorization: "Bearer " + params.token,
        },
        data: params.body,
    })
        .then((data) => ({
        id: data.data.id,
        name: data.data.title,
    }))
        .catch((error) => {
        throw new Error("Error: Update File Endpoint");
    });
};
exports.updateFileData = updateFileData;
const getUserIdFromAccessToken = async (params) => {
    return await (0, axios_1.default)({
        method: "get",
        url: `https://www.googleapis.com/oauth2/v1/tokeninfo?access_token=${params.token}`,
    })
        .then((data) => {
        return data["data"]["user_id"];
    })
        .catch((error) => {
        throw new Error("Error: Get User Id from endpoint Endpoint");
    });
};
exports.getUserIdFromAccessToken = getUserIdFromAccessToken;
const getFileData = async (params) => {
    return await (0, axios_1.default)({
        method: "get",
        url: `https://www.googleapis.com/drive/v2/files/${params.fileId}?alt=media`,
        headers: {
            Accept: "application/json",
            Authorization: "Bearer " + params.token,
        },
    })
        .then((data) => {
        return { data: data.data };
    })
        .catch((error) => {
        throw new Error("Error: Get File Endpoint");
    });
};
exports.getFileData = getFileData;
